File: node/src/finance.ts
import { HttpClient } from './http-client';
import type { ShipzyConfig } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export interface Invoice {
    id: string;
    invoice_number: string;
    amount: number;
    currency: string;
    status: 'paid' | 'pending' | 'failed' | 'refunded';
    description: string;
    created_at: string;
    paid_at?: string;
    download_url?: string;
}

export interface Subscription {
    id: string;
    status: string;
    plan: string;
    price: number;
    currency: string;
    start_date: string;
    next_billing_date: string;
    cancel_at_period_end: boolean;
    created_at: string;
}

// ============ Client ============

export class FinanceClient extends HttpClient {
    async getInvoices(): Promise<ApiResult<Invoice[]>> {
        return this.request('/api/finance/invoices');
    }

    async listSubscriptions(): Promise<ApiResult<Subscription[]>> {
        return this.request('/api/finance/subscriptions');
    }

    async cancelSubscription(id: string): Promise<ApiResult<{ status: string }>> {
        return this.request(`/api/finance/subscriptions/${encodeURIComponent(id)}/cancel`, 'POST');
    }
}

File: python/zymeup_sdk/finance.py
"""Zymeup Python SDK - Finance Client"""
from typing import Optional, Dict, Any, List


class FinanceClient:
    """Finance and subscription client"""

    def __init__(self, client):
        self._client = client

    def get_invoices(self) -> Dict[str, Any]:
        """List invoices"""
        return self._client.request("GET", "/api/v1/invoices")

    def list_subscriptions(self) -> Dict[str, Any]:
        """List subscriptions"""
        return self._client.request("GET", "/api/v1/subscriptions")

    def cancel_subscription(self, subscription_id: str) -> Dict[str, Any]:
        """Cancel a subscription"""
        return self._client.request("POST", f"/api/v1/subscriptions/{subscription_id}/cancel")

    def restore_subscription(self, subscription_id: str) -> Dict[str, Any]:
        """Restore a cancelled subscription"""
        return self._client.request("POST", f"/api/v1/subscriptions/{subscription_id}/resume")

    def download_invoice(self, invoice_id: str) -> Dict[str, Any]:
        """Download an invoice"""
        return self._client.request("GET", f"/api/v1/invoices/{invoice_id}/download")
